### DISCUSSION TITLE
Long term planning (year 2027 onwards)

### DISCUSSION THREAD

**Discussion by olljanat** [burmilla/os#186] [Polls]
:warning::warning::warning: [v2.0.5](https://github.com/burmilla/os/releases/tag/v2.0.5) is potentially last version of BurmillaOS. :warning::warning::warning:

Please, check discussion about potential replacement in [here](https://github.com/burmilla/os/discussions/190)

---

It's almost been four years since first Burmilla OS version [1.9.0](https://github.com/burmilla/os/releases/tag/v1.9.0) was released.
Most important driver for that was allow ex Rancher OS users to update their systems to versions which are still receiving security updates and we have got some new users too.

From GitHub statistics I can see that v2.0.0 ISO file have been downloaded over 600 and v2.0.1 over 500 times.
On top of that many users upgraded from earlier version/Rancher OS which is not visible on those numbers.

What worries me a bit is that Rancher OS v1.5.8 ISO file is still downloaded almost everyday. Docker and Linux kernel versions in it contains known security vulnerabilities so I can only hope that those are non-production systems.

Main components in Burmilla OS 2.0.x versions are:
* Linux kernel 5.10.x which will run of out support in [December 2026](https://www.kernel.org/category/releases.html)
* Debian 11 "Bullseye" based console which will run of out support in [August 2026](https://wiki.debian.org/LTS)

Debian 12 "Bookworm" is using Linux kernel 6.1.x and is supported until June 2028 so most likely that kernel version will be supported also until that same date.

However, before starting implementing new version with those components I would like to confirmation that there is enough of those who are running production workloads on top of Burmilla OS. So please answer to poll below.

**Comment by CerberusZ** [comment]
This is an amazing project and I really hope you continue supporting it.

**Reply by olljanat** [reply]
> This is an amazing project

True but that does not mean that some other distro cannot be equally good or even better.
Opened discussion to #190 about one alternative.

**Reply by stevecharon** [reply]
tried flatcar for another project but it was more unflexible for enterprise things like backup agents for this or that. 
rancheros/burmillaos was giving me less headaches. 
and I like the smaller footprint of these images. but an more up-to-date kernel is crucial for a broad support.
with the debian or ubuntu console I was almost always able to get it to work.
Flatcar with its read-only partition was a nuisance in that regard.
updating flatcar was ok.
Now moving onto standard ubuntu boxes for docker hosts due to better acceptance by my colleagues and the easier installation of said agents and tools.
Still using burmillaos for a few projects though.

**Comment by benok** [comment]
As v2.0.2 is released, I think this vote should be extended. 
Only a few beta users voted.

**Reply by olljanat** [reply]
Keeping mind those schedules mentioned in first message, I think that this voting can be considered to be open somewhere like June 2026 so there is plenty of time also those who do not update that often to see info message.

And people still can report issues or create feature requests and I will check those case by case.

**Comment by olljanat** [comment]
Because main challenge here is kernel, one alternative which I have been thinking lately would be using kernel binaries from Debian.

However, my first [merge request](https://salsa.debian.org/images-team/debian-cd/-/merge_requests/45) to Debian isn’t looking too promising at the moment so fixing bugs might be tricky if we found any.

**Comment by pwFoo** [comment]
It would be easy to move to os like rocky, alma or similiar... but I loved to use such a special docker os like rancheros / burmillaos. Running and updating it since one early rancheros release...

Played around with linuxkit base to build an minimal docker os myself, but some important parts like os updates, ssh to host, ... are tricky.
Anyone building / playing around with such a fun os build? Would be interested to take a look into...

**Reply by olljanat** [reply]
Yeah well BurmillaOS maintenance also quite easy and there is no that much to do after major release is ready and published. It is just these larger jumps  in kernel versions and potentially needed refactoring for existing code (which are explained in open issues) which needs extra effort and would benefit from contributions…

**Reply by olljanat** [reply]
> some important parts like os updates, ssh to host, ... are tricky

You might want to try new prototype in [here](https://github.com/burmilla/os-elemental).

It is based on framework which Rancher/SUSE uses nowadays to build minimal disk images for Kubernetes. Those are based on OpenSUSE but they made framework generic enough that it can be used with any distro.

So my thinking is that BurmillaOS v3 would be simply repo which automatically build and publish new disk image when new Docker version is released (based on Debian or Ubuntu) and those who needs customization can then simply fork that repo (or just use cloud-init if that is enough for you).

**Reply by olljanat** [reply]
> You might want to try new prototype in [here](https://github.com/burmilla/os-elemental).

I wonder that how many tested this? Don’t you think that it solves the same problem than RancherOS/BurmillaOS?

Last couple of months I have been building [Nomad QEMU node](https://github.com/olljanat/nomad-qemu-node) which is also built with Elemental Toolkit and I find it very productive way to build (near) maintenance free distros for different purposes.

And it would be very simple to make that and BurmillaOS v3 (if based on elemental toolkit) to share 95% of the code.

**Comment by olljanat** [comment]
I see that someone is running Nomad in production with BurmillaOS.

Can I ask a bit more details about that setup? Example to #6 or as separate issue/discussion

**Reply by mdef** [reply]
@olljanat We use burmilla as a docker server -  for local services (docker-compose) and kubernetes which is also running as docker container. 
We manage  ~10 such installations and burmilla has been drop-in replacement for rancher. where Rancher was replacement for CoreOS.
Still hope for future releases.

**Reply by olljanat** [reply]
@mdef Can I ask that how you run Kubernetes as docker containers?

> Still hope for future releases.

I know that many is hoping same but it is getting harder because Docker switched to faster release cycle and there is more breaking changes now and more like https://github.com/burmilla/os/issues/195#issuecomment-3536049578 is coming which would need major changes to BurmillaOS.

However, I now created [new automation](https://github.com/burmilla/docker-ce-packaging) which allow us build Docker 25.0.x binaries and added [Docker v25.0.14 to BurmillaOS](https://github.com/burmilla/os-services/pull/55) which might be optimal choice for those who just want stable environment to run containers and do not need new features (and I still took docker compose and buildx from Docker 29.0.1 so new features included to those should still works).

Would be interesting to hear if that works for your and others use cases? Because if that does then I potentially would prefer to switch BurmillaOS defaulting to those Docker 25.0.x versions in next release.

**Reply by mdef** [reply]
@olljanat I'm using official [compose from k3s](https://github.com/k3s-io/k3s/blob/main/docker-compose.yml)

**Comment by prologic** [comment]
Can we perhaps potentially think of it handing over maintenance of this project as I still actively use BurmilaOS so I would want to see its existence continued!

**Reply by olljanat** [reply]
@prologic Absolutely. I have been hoping that someone will do that offer.

Most logical point to do hand over would be version 3.0.0. I can still help with that but you would take lead of it.
Created [project](https://github.com/orgs/burmilla/projects/1/views/1) of it and basically you can start working on those by creating pull requests against of master branch in relevant repos.

Items in "Todo" column are ones which at least would need to be included.
Items in "No Status" can be considered but they are not mandatory to keep this project alive.

**Reply by prologic** [reply]
Sorry for belated reply, xmas and all that 🤣 Reaching out to the rest of the community here. I can't do this alone, I'd like someone to co-maintain this project with if anyone else is interested?

I'm also thinking based on the votes above, that we'd descope the project a bit and drop support for Kubernetes. Most of the use-cases seems to be around standalone Docker and Docker Swarm, which also match my use-case (_although I run production Swarm clusters this way_).

**Reply by olljanat** [reply]
> I'm also thinking based on the votes above, that we'd descope the project a bit and drop support for Kubernetes.

There is some discussion about topic in #47 but basically Kubernetes never have been officially supported. Also the whole reason why Rancher deprecated RancherOS is that it was built before Kubernetes became popular.

> I can't do this alone, I'd like someone to co-maintain this project with if anyone else is interested?

Yes, that is the whole reason why this discussion exists. None of the open source projects cannot live in longterm unless there is multiple maintainers.

However this is one of the easiest projects out there because very little needs to be done between major versions.

**Reply by prologic** [reply]
Are you the current and primary maintainer? Perhaps you/i can combine forces and continue maintaining BurmillaOS?

**Reply by mdef** [reply]
> Are you the current and primary maintainer? Perhaps you/i can combine forces and continue maintaining BurmillaOS?

I would join if no objections.

**Reply by psychowood** [reply]
I'm just wondering if the BurmillaOS (which I'm still using from my homelab btw, deployed years ago) could be made completely stateless without the need for installing it at all.  

Something like an .iso distribution that looks for the configuration/init files in specific path patterns (by disk label/filesystem/directory/whatever), and would load a ros config that should include the root for /var/lib/docker and so on.

At that point an upgrade/rollback could be just an image swap, and also customization could be done via elemental toolkit even with an online image builder of sort.

If you believe this could make sense I could try to experiment a bit with it.

**Reply by mdef** [reply]
I believe this is off topic for the thread, but 

https://burmillaos.org/docs/installation/boot-process/cloud-init/
Would not cover your needs?

> I'm just wondering if the BurmillaOS (which I'm still using from my homelab btw, deployed years ago) could be made completely stateless without the need for installing it at all.  
> 
> Something like an .iso distribution that looks for the configuration/init files in specific path patterns (by disk label/filesystem/directory/whatever), and would load a ros config that should include the root for /var/lib/docker and so on.
> 
> At that point an upgrade/rollback could be just an image swap, and also customization could be done via elemental toolkit even with an online image builder of sort.
> 
> If you believe this could make sense I could try to experiment a bit with it.

**Reply by psychowood** [reply]
I thought it could be considering a long term planning scenario. I admit I havent dug too much into cloud-init because [from the docs](https://burmillaos.org/docs/configuration/#cloud-config) I understood that it was just for automating the first installation.

`Typically, when you first boot the server, you pass in a cloud-config file to configure the initialization of the server. After the first boot, if you have any changes for the configuration, it’s recommended that you use ros config to set the necessary configuration properties. Any changes will be saved on disk and a reboot will be required for changes to be applied.`

Sorry if OT, in case I'll move to another discussion after investigating more. 👍

**Reply by olljanat** [reply]
> Are you the current and primary maintainer?

@prologic  Yes, I forked BurmillaOS from RancherOS about 5 years ago and have done most the maintaining since then. Some point there was some others but haven’t seen them here for a while.

> Perhaps you/i can combine forces and continue maintaining BurmillaOS?

Sure, I’m happy to help and like mentioned earlier my plan is to continue maintaining v2.0.x but it would be nice to find new leader/main maintainer for 3.x.

@psychowood please create new [issue](https://github.com/burmilla/os/issues) about your idea including what you are trying to achieve with that? Also sounds than you might want to try install non-latest BurmillaOS and upgrade it to latest so you see how switching between those currently works. I have done a lot of research about different options and can explain in that issue what  your idea would require in practice and what would be pros and cons in that.

**Reply by prologic** [reply]
Yeah okay. I'm more than happy to take lead on this project. What do you need me to do over the next coming weeks? 🤔 I will need to spend some time actually understanding the codebase, structure and how everything fits together. Since the RancherOS days, and now BurmillaOS, I've only really been a production user (_yes, running Swarm clusters_).

**Reply by olljanat** [reply]
> What do you need me to do over the next coming weeks? 🤔

Look https://github.com/burmilla/os/discussions/186#discussioncomment-14981140
First thing would be going through those which needs to be done, write any open questions to there and agree that who actually will handle each of those.

> Since the RancherOS days, and now BurmillaOS, I've only really been a production user (_yes, running Swarm clusters_).

Relating to Swarm, one important thing is to make sure that we do not break compatibility with Docker 25.x versions example when handling https://github.com/burmilla/os/issues/212

Newer Docker versions haven’t been tested with Swarm by Mirantis and there is some nasty bugs which why I recently added https://github.com/burmilla/os-services/pull/55

**Reply by olljanat** [reply]
@prologic Are you still working on this/planning to do so?

I have been recently doing same than everyone else, playing with Claude Code and yesterday I asked it to fix issues in BurmillaOS documentation where it did very good job compared that I did just give very short prompt with high level instructions for it: https://github.com/burmilla/burmilla.github.io/pull/41

So getting those needed updates done should be anymore issue anymore except that someone has available tokens to use for this purpose and time to validate those results.

It would be good if we can find someone with OpenAI Codex and Gemini Pro so those can be used to cross check each others work. It is also very good way to learn for anyone who is working with IT, even when you are not developer (I certainly isn't).

PS. Security of these AI agents is big challenge so I'm at the momement working on my [Firewall4AI](https://github.com/olljanat-ai/Firewall4AI) project which targets to provide better starting point for those who wants to evaluate them. Funny fact in that is that now Claude Code is building it's own prison...

**Reply by prologic** [reply]
Yup!

**Reply by olljanat** [reply]
@prologic Ok, good. Added triagle permissions for you.

Please, self assign those issues you are actively working on to yourself so we don't waste time by working on same things.

**Comment by ahmedmunir1179** [comment]
What about other binaries and libs like use etc. try also update them along with kernel, great work. Happy to see this fork of rancheros

**Reply by olljanat** [reply]
Replied to other discussion you created https://github.com/burmilla/os/discussions/213#

**Comment by olljanat** [comment]
With big help from Claude, I managed to build and release [v3.0.0-beta1](https://github.com/burmilla/os/releases/tag/v3.0.0-beta1).
Please, test it and report bugs to #226

Closing this discussion as it got solved on this way.
Thanks a lot of all feedback.
